services/collectd/service.go

```diff
@@ -336,32 +358,49 @@ func (s *Service) serve() {
 	//   Ethernet).
 	buffer := make([]byte, 1452)
 
+	type ReadPayload struct {
+		n   int   // bytes read by ReadFromUDP
+		err error // any error returend by ReadFromUDP
+	}
+
+	rp := ReadPayload{}
+
+	reads := make(chan ReadPayload)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				rp.n, _, rp.err = s.conn.ReadFromUDP(buffer)
```

Comment:
I spent five minutes looking this code because initially it looked racy to me. Two goroutines accessing the same value unsynchronised, e.g., `rp.n`. Because `ReadPayload.n` is copied when it's passed to the channel there isn't a race, but it's a bit nuanced. Maybe worth a comment for future folks looking at this code?

Reply:
agreed.

Reply:
oh -- that's a bug.  `rp` should be `cur` in some parts.  will fix.

Reply:
This is racy on `buffer` as there are multiple concurrent accesses with one being a write.

![image](https://user-images.githubusercontent.com/52852/75065124-73f51500-54a5-11ea-8bc2-e94693396953.png)

Reply:
Should have just gone with my gut heh